lls/cache.c

```diff
@@ -0,0 +1,481 @@
+			"lls: allocation for request of type %d for %s failed",
+			ty, ip_str == NULL ? "<unknown>" : ip_str);
+	}
+
+	req->ty = ty;
+	req->cache = cache;
+	memcpy(req->ip_be, ip_be, cache->key_len);
+
+	switch (ty) {
+	case LLS_REQ_RESOLVE:
+		/* FALLTHROUGH */
+	case LLS_REQ_SUBSCRIBE:
+		req->u.hold = *(struct lls_hold *)req_arg;
+		break;
+	case LLS_REQ_PUT:
+		req->u.put = *(struct lls_put *)req_arg;
+		break;
+	case LLS_REQ_MODIFY:
+		req->u.mod = *(struct lls_mod *)req_arg;
+		break;
+	default:
+		mb_free_entry(&lls_conf->requests, req);
+		ip_str = cache->ip_str(cache, ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: unknown request type %d for %s failed",
+			ty, ip_str == NULL ? "<unknown>" : ip_str);
+		return -1;
+	}
+
+	ret = mb_send_entry(&lls_conf->requests, req);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static void
+lls_cache_del(struct lls_cache *cache, struct lls_record *record)
+{
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+
+	int32_t ret = rte_hash_del_key(cache->hash, record->map.ip_be);
+	if (unlikely(ret == -ENOENT)) {
+		ip_str = cache->ip_str(cache, record->map.ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"No map found, record for %s not deleted\n",
+			ip_str == NULL ? cache->name : ip_str);
+			return;
+	} else if (unlikely(ret == -EINVAL)) {
+		ip_str = cache->ip_str(cache, record->map.ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, record for %s not deleted\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+}
+
+void
+lls_cache_scan(struct lls_cache *cache, struct gatekeeper_if *front,
+	struct gatekeeper_if *back)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	uint8_t *ip_be;
+	void *data;
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+	struct lls_config *lls_conf = get_lls_conf();
+	time_t now = time(NULL);
+
+	index = rte_hash_iterate(cache->hash, (void *)&ip_be, &data, &iter);
+	while (index >= 0) {
+		struct lls_record *record = &cache->records[index];
+		uint32_t timeout;
+
+		/*
+		 * If a map is already stale, continue to
+		 * try to resolve it while there's interest.
+		 */
+		if (record->map.stale) {
+			if (record->num_holds > 0)
+				send_request(lls_conf, cache, ip_be, NULL);
+			else
+				lls_cache_del(cache, record);
+			goto next;
+		}
+
+		if (record->map.port_id == front->id)
+			timeout = cache->front_timeout_sec;
+		else if (back != NULL && record->map.port_id == back->id)
+			timeout = cache->back_timeout_sec;
+		else {
+			ip_str = cache->ip_str(cache, ip_be, ip_buf,
+				cache->key_str_len);
+			RTE_LOG(ERR, GATEKEEPER,
+				"lls: map for %s has an invalid port %hhu\n",
+				ip_str == NULL ? cache->name : ip_str,
+				record->map.port_id);
+			goto next;
+		}
+
+		if (now - record->ts > timeout) {
+			record->map.stale = true;
+			update_subscribers(record, false);
+			if (record->num_holds > 0)
+				send_request(lls_conf, cache, ip_be,
+					&record->map.ha);
+		}
```

Comment:
If we only send a unicast probe once the entry has expired, we may cause disruptions for addresses that are always on and always being used since that for a short period the entry will be stale. My suggestion is to have an `else if` here to send unicast probes when `timeout > LLS_CACHE_SCAN_INTERVAL && now - record->ts > timeout - LLS_CACHE_SCAN_INTERVAL`.